File: scratch/purust-fbip-20260916/benchmark/src/Test/PolymorphismFFI.cc
#include "purescript.h"
#include <functional>

namespace {
    using IntFn = std::function<int(int)>;
    struct Monoidish {
        int mempty;
        std::function<IntFn(int)> mappend;
    };

    int polyLoop(const Monoidish& dict, int n, int acc) {
        while (n > 0) {
            acc = dict.mappend(acc)(dict.mempty);
            --n;
        }
        return acc;
    }
}

FOREIGN_BEGIN(Test_PolymorphismFFI)
exports["runPolymorphismFFI"] = [](const boxed& in) -> boxed {
    const Monoidish dict{1, [](int x) -> IntFn {
        return [x](int y) { return x + y; };
    }};
    const int result = polyLoop(dict, unbox<int>(in), 0);
    return result;
};
FOREIGN_END

File: scratch/purust-fbip-20260916/benchmark/src/Test/PolymorphismFFI.php
<?php
$exports['runPolymorphismFFI'] = function($limit) {
    $n = (int)$limit;
    $dict = (object)[
        "mempty_" => 1,
        "mappend_" => function($x) {
            return function($y) use ($x) {
                return $x + $y;
            };
        }
    ];
    $acc = 0;
    while ($n > 0) {
        $acc = ($dict->mappend_)($acc)($dict->mempty_);
        $n--;
    }
    return $acc;
};
return $exports;
